libazureinit/src/status.rs

```diff
@@ -0,0 +1,337 @@
+    let path = custom_path.unwrap_or("/sys/class/dmi/id/product_uuid");
+
+    let system_uuid = match fs::read_to_string(path) {
+        Ok(s) => s.trim().to_lowercase(),
+        Err(err) => {
+            tracing::error!("Failed to read VM ID from {}: {}", path, err);
+            return None;
+        }
+    };
+
+    if system_uuid.is_empty() {
+        tracing::info!(target: "libazureinit::status::retrieved_vm_id", "VM ID file is empty at path: {}", path);
+        return None;
+    }
+
+    if is_vm_gen1(mock_efi_path) {
+        match Uuid::parse_str(&system_uuid) {
+            Ok(uuid_parsed) => {
+                let swapped_uuid =
+                    swap_uuid_to_little_endian(*uuid_parsed.as_bytes());
+                let final_id = swapped_uuid.to_string();
+                tracing::info!(target: "libazureinit::status::retrieved_vm_id", "VM ID (Gen1, swapped): {}", final_id);
+                Some(final_id)
+            }
+            Err(e) => {
+                tracing::error!(
+                    "Failed to parse system UUID '{}': {}",
+                    system_uuid,
+                    e
+                );
+                // fallback to the raw string
+                Some(system_uuid)
+            }
+        }
+    } else {
+        tracing::info!(target: "libazureinit::status::retrieved_vm_id", "VM ID (Gen2, no swap): {}", system_uuid);
+        Some(system_uuid)
+    }
+}
+
+/// This function checks whether a provisioning status file exists for the current VM ID.
+/// If the file exists, provisioning has already been completed and should be skipped.
+/// If the file does not exist or the VM ID has changed, provisioning should proceed.
+///
+/// - Returns `true` if provisioning is complete (i.e., provisioning file exists).
+/// - Returns `false` if provisioning has not been completed (i.e., no provisioning file exists).
+pub fn is_provisioning_complete(
+    config: Option<&Config>,
+    vm_id: Option<String>,
+) -> bool {
+    let vm_id = vm_id.or_else(|| get_vm_id(None, None));
+
+    if let Some(vm_id) = vm_id {
+        let file_path =
+            get_provisioning_dir(config).join(format!("{}.provisioned", vm_id));
+        if std::path::Path::new(&file_path).exists() {
+            tracing::info!("Provisioning already complete. Skipping...");
+            return true;
+        }
+    }
+    tracing::info!("Provisioning required.");
+    false
+}
+
+/// This function creates an empty file named after the current VM ID in the
+/// provisioning directory. The presence of this file signals that provisioning
+/// has been successfully completed.
+pub fn mark_provisioning_complete(
+    config: Option<&Config>,
+    vm_id: Option<String>,
+) -> Result<(), Error> {
+    check_provision_dir(config)?;
+
+    let vm_id = vm_id.or_else(|| get_vm_id(None, None));
+
+    if let Some(vm_id) = vm_id {
+        let file_path =
+            get_provisioning_dir(config).join(format!("{}.provisioned", vm_id));
```

Comment:
at some point we'd possibly want a similar marker file for pre-preprovisioned. I just wonder if it makes more sense to have state information in one file or to do these markers